Add a horizontal auto scaler and Metrics

l0nax opened the pull request:
This PR adds a Horizontal Auto-Scaler and Metrics to the `kubenab` Application.

The HPA (Horizontal Pod Auto-Scaler) can be configured to work with native K8s Metrics (_CPU Utilization_, etc.) or with Prometheus Metrics (_Request per Second_; _Request Duration_; etc.)

Closes #11 and Closes #13


## TODO

- [x] Add Prometheus Metrics
- [x] Use the CRD's of the Prometheus-Operator to collect the Metrics
- [x] Add Metrics to Helm Chart
- [x] Document usable Metrics of _HPA_ in the Helm-Chart `values.yaml`
  * [x] Done but only in the `values.yaml` ==> _Needs Works_
- [x] Document Usage of the _HPA_

## Changelog

Here is a _Changelog_, which can be later used when releasing a new version.


- **Helm-Chart:** Add HPA with build-in and custom Metrics
- **Helm-Chart:** Add Prometheus Metrics Scraper
- **Helm-Chart:** Add Application Metrics
- **Raw Deployments:** Update `apiVersion` to support deployment on a 1.16+ K8s-Cluster

l0nax committed:
Update `go.mod` and `go.sum`

Update/ Sort dependencies and add new Required Dependecies.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Add two Prometheus Metrics

* http_request_duration_milliseconds
        Duration of all request in ms.
        Distinguish between 'mutate' and 'validate'.

* http_ping_requests_total
        Total '/ping' Requests

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Register Prometheus Metrics

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Add '/metrics' Endpoint

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Add Request Time Logging to '/mutate' and '/validate'

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Rename 'http_ping_requests_total' Prometheus Metric

Rename the Prometheus Metric 'http_ping_requests_total' to
'http_requests_total', to prepare that also the Metrics from the other
Endpoints can be tracked.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Add label to 'http_requests_total' Metric

Add `api_endpoint` label to the 'http_requests_total' Metric, to
distinguish between the different API-Endpoints.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Add 'http_request_total' Metric to all Endpoints

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Add Helm Values for Prometheus Deployment

Add tree new Values to the `values.yaml` Chart which provides the
possibilitie to configure if a Prometheus Configuration (ServiceMonitor,
Prometheus) should be created to monitor the Metrics of _kubenab_.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Add Metrics Template to Helm-Chart

Implements the Kuebernetes Manifests to monitor the _kubenab_ Metrics.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

pull[bot] committed:
Merge pull request #1 from jfrog/master

[pull] master from jfrog:master

rimusz commented:
@l0nax thanks for your PRs 👍 
Could you please rebase this PR from master, as I did split `main.go` to two files:
- main.go
- admissions.go

Thanks

pull[bot] committed:
Merge pull request #2 from jfrog/master

[pull] master from jfrog:master

l0nax commented:
@rimusz `master` rebased over `feature/11_add-a-horizontal-auto-scaler`

l0nax committed:
Merge Changes from PR #15

Since the split of _main.go_ into _main.go_ and _admissions.go_ in
Commit 542d1a04b5e282336159b78217f77d45ee2d648c changes made in this
Branch where not included, so this Patch applies those Changes.

Also the `go.sum` was re-generated.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

rimusz commented:
Please move HPA code out from the master and put to `addmissions.go`, thanks

l0nax commented:
Done

l0nax committed:
Remove duplicated Code Line

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Update 'apiVersion' in K8s Deployments

With the realse of Kubernetes v1.16 there where several Deprecation made
to the API[^1], so `extensions/v1beta1` isn't usable easily anymore in a
K8s-Cluster running version 1.16+.

[^1]: https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.16.md#deprecations-and-removals

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Add min Kubernetes Version to Helm-Chart

As mentoined in Commit 2a1a76d259eaa23a06a2de18d26483c8b4f8b32a with the
Release of Kubernetes v1.16 there where several APIs marked as
deprecated.
To prevent failed Deployments we define a minimum Kubernetes version
1.16.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Add VIM Swap Files to .gitignore

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Add 'hpa' Values to Helm-Chart

Add 'hpa.enabled', 'hpa.minReplicas' and 'hpa.maxReplicas' to the
default Helm-Chart 'values.yaml'.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Add 'cpuUtilization' Option to HPA Helm-Chart

Add 'cpuUtilization' to 'hpa.metrics' in the default Helm-Chart
'values.yaml'.

'cpuUtilization' is a build-in Metrics from Kubernetes, so Users don't
have to deploy Prometheus(-Operator) to use the HPA Feature.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Add Common Labels Helper to Helm-Chart

Define 'kubenab.labels' to prevent duplicated Code (of labels) over all
the Chart Templates.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Replace hard-coded Labels with defined Value

Replace all hard-coded Labels in the Helm-Chart Templates with the,
in Commit cc3f288019d7d3235a62d6876d0ba22246270d93 defined `kubenab.labels`.
This will prevent from having duplicated Code in the Chart Templates
directory. So it will be easier to maintain.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Add 'httpRequests' Metric to HPA in Helm-Chart

Add new Metric 'httpRequests' that can be enabled in the Helm-Chart.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Add 'latency' Metric to HPA in Helm-Chart

Add new Metric 'latency' that can be enabled in the Helm-Chart.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Add required ConfigMap Changes in NOTES.txt

The User must manually edit the ConfigMap of the prometheus-adapter to
enable the advanched Metrics.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Add HPA Template to Helm-Chart

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax commented:
But then we would have Problems with Deployments on 1.16+ K8s-Clusters.
Because the APIs has been marked as deprecated.

https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.16.md#deprecations-and-removals

l0nax commented:
We could add a check for the Cluster Version and then change the `apiVersion` dynamically.

rimusz commented:
yes, the ApiVersion needs to be changed dynamically as e.g. cloud providers there is no k8s 1.16 support yet and it is very rear to see k8s 1.16 in prod yet

l0nax committed:
Reduce minimum K8s Version to '1.12' in Helm-Chart

As discussed in PR #12 (Comments), the minimum Kubernetes Version has to
be 1.12.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Add check for APIVersions to Helm-Chart Templates

Add checks if the target Kubernetes Cluster has/ provides a specific
apiVersion to provide support of "<=1.12" and ">=1.16"
Kubernetes-Clusters.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax commented:
Added Support for K8s-Clusters with version _<=1.12_ and _>=1.16_

l0nax committed:
Correct Value Name Spelling Misstake in Helm-Chart

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Update `http_requests_total` Metric Data

Update Help Description and remove some unused Labels of the
`http_requests_total` Metric.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Change `http_request_duration_milliseconds` Metric Type

Change the Metric type of `http_request_duration_milliseconds` from
Summary to Histogram after some Research because a Histogram is better
for the avg Response Time/ Latency.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Update Prometheus CRD Version

Update the Prometheus Version of the Prometheus CRD to v2.5.0 to support
some new Features and enhancements.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Update Insecure TLS Flag of ServiceMontior

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Short Metric Labelnames in NOTES.txt

Short the used Metric labelsnames to prevent missconfiguration and
confiusion.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Fix Prometheus Ops Struct Field misspelling

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax commented:
The custom Metric for `Request per Second` in combination with the HPA is implemented and works but the `Request Duration/ Latency` seems to be difficult to implement with the `prometheus-adapter` I think I will need some more time.

Sorry :/

rimusz commented:
take your time @l0nax and thanks for your contributions 👍

l0nax committed:
Update HTTP Latency Metric Adapter Query

The `NOTES.txt` Template will now print a full working Entry for the
K8s-Prometheus-Adapter ConfigMap to get the HTTP Lantecy.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Update 'Request/s' Adapter Config

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax commented:
The K8s-Prometheus-Adapter Config now finally works:

ConfigMap entries:
```yaml
  config.yaml: |
    rules:
    - seriesQuery: '{namespace!="",pod!=""}'
      name:
        matches: "^(.*)_milliseconds_bucket"
        as: "http_request_duration"
      resources:
        overrides:
          namespace: {resource: "namespace"}
          pod: {resource: "pod"}
      metricsQuery: 'histogram_quantile(0.99, sum(rate(<<.Series>>[5m])) by (le, api_method, endpoint, <<.GroupBy>>))'
    - seriesQuery: 'http_requests_total{namespace!="",pod!=""}'
      resources:
        overrides:
          namespace: {resource: "namespace"}
          pod: {resource: "pod"}
      name:
        matches: "^(.*)_total"
        as: "${1}_per_second"
      metricsQuery: 'sum(rate(<<.Series>>{<<.LabelMatchers>>}[2m])) by (<<.GroupBy>>)'
```


### Raw API Check

_NOTE: To get "realistic" results in the `custom.metrics` API I run the `benchmark.sh` Script in the background to generate load_

#### Available Metrics

* Command: `kubectl get --raw "/apis/custom.metrics.k8s.io/v1beta1" | jq`

```json
{
  "kind": "APIResourceList",
  "apiVersion": "v1",
  "groupVersion": "custom.metrics.k8s.io/v1beta1",
  "resources": [
    {
      "name": "pods/http_request_duration",
      "singularName": "",
      "namespaced": true,
      "kind": "MetricValueList",
      "verbs": [
        "get"
      ]
    },
    {
      "name": "pods/http_requests_per_second",
      "singularName": "",
      "namespaced": true,
      "kind": "MetricValueList",
      "verbs": [
        "get"
      ]
    },
    {
      "name": "namespaces/http_requests_per_second",
      "singularName": "",
      "namespaced": false,
      "kind": "MetricValueList",
      "verbs": [
        "get"
      ]
    },
    {
      "name": "namespaces/http_request_duration",
      "singularName": "",
      "namespaced": false,
      "kind": "MetricValueList",
      "verbs": [
        "get"
      ]
    }
  ]
}
```

#### Requests per second

* Command: `kubectl get --raw "/apis/custom.metrics.k8s.io/v1beta1/namespaces/monitoring/pods/*/http_requests_per_second" | jq`

```json
{
  "kind": "MetricValueList",
  "apiVersion": "custom.metrics.k8s.io/v1beta1",
  "metadata": {
    "selfLink": "/apis/custom.metrics.k8s.io/v1beta1/namespaces/monitoring/pods/%2A/http_requests_per_second"
  },
  "items": [
    {
      "describedObject": {
        "kind": "Pod",
        "namespace": "monitoring",
        "name": "kubenab-deployment-6468d895f5-kdvkl",
        "apiVersion": "/v1"
      },
      "metricName": "http_requests_per_second",
      "timestamp": "2019-10-09T19:17:07Z",
      "value": "340767m"
    }
  ]
}
```

#### Response latency

* Command: `kubectl get --raw "/apis/custom.metrics.k8s.io/v1beta1/namespaces/monitoring/pods/*/http_request_duration" | jq`

```json
{
  "kind": "MetricValueList",
  "apiVersion": "custom.metrics.k8s.io/v1beta1",
  "metadata": {
    "selfLink": "/apis/custom.metrics.k8s.io/v1beta1/namespaces/monitoring/pods/%2A/http_request_duration"
  },
  "items": [
    {
      "describedObject": {
        "kind": "Pod",
        "namespace": "monitoring",
        "name": "kubenab-deployment-6468d895f5-kdvkl",
        "apiVersion": "/v1"
      },
      "metricName": "http_request_duration",
      "timestamp": "2019-10-09T19:16:56Z",
      "value": "4952m"
    }
  ]
}
```

l0nax committed:
Correct Metric Names in HPA template

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Add default latency Metric avg to values.yaml

Add a default a average Metric Value for the `latency` Metric in
values.yaml of the Helm-Chart.

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax committed:
Use v2beta2 API Version of HPA

Signed-off-by: Emanuel Bennici <benniciemanuel78@gmail.com>

l0nax commented:
The _HPA_ Feature works now perfectlly.

The _HPA_ reacted after 10 seconds.

```bash
~] k describe hpa release-name-kubenab
Name:                                  release-name-kubenab
Namespace:                             monitoring
Labels:                                app.kubernetes.io/instance=release-name
                                       app.kubernetes.io/managed-by=Tiller
                                       app.kubernetes.io/name=kubenab
                                       app.kubernetes.io/version=0.3.2
                                       helm.sh/chart=kubenab-0.0.7
Annotations:                           kubectl.kubernetes.io/last-applied-configuration:
                                         {"apiVersion":"autoscaling/v2beta2","kind":"HorizontalPodAutoscaler","metadata":{"annotations":{},"labels":{"app.kubernetes.io/instance":"...
CreationTimestamp:                     Wed, 09 Oct 2019 21:48:42 +0200
Reference:                             Deployment/kubenab-deployment
Metrics:                               ( current / target )
  "http_requests_per_second" on pods:  568870m / 300
  "http_request_duration" on pods:     4952m / 5
Min replicas:                          1
Max replicas:                          10
Deployment pods:                       1 current / 2 desired
Conditions:
  Type            Status  Reason              Message
  ----            ------  ------              -------
  AbleToScale     True    SucceededRescale    the HPA controller was able to update the target scale to 2
  ScalingActive   True    ValidMetricFound    the HPA was able to successfully calculate a replica count from pods metric http_requests_per_second
  ScalingLimited  False   DesiredWithinRange  the desired count is within the acceptable range
Events:
  Type    Reason             Age   From                       Message
  ----    ------             ----  ----                       -------
  Normal  SuccessfulRescale  15s   horizontal-pod-autoscaler  New size: 2; reason: pods metric http_requests_per_second above target

```

The `deployment` get's scaled up:
```
[...]
Conditions:
  Type           Status  Reason
  ----           ------  ------
  Progressing    True    NewReplicaSetAvailable
  Available      True    MinimumReplicasAvailable
OldReplicaSets:  <none>
NewReplicaSet:   kubenab-deployment-6468d895f5 (2/2 replicas created)
Events:
  Type    Reason             Age   From                   Message
  ----    ------             ----  ----                   -------
  Normal  ScalingReplicaSet  97s   deployment-controller  Scaled up replica set kubenab-deployment-6468d895f5 to 2
```

l0nax commented:
@rimusz The PR is now ready to be reviewed.

l0nax commented:
/label enhancement

elioengcomp commented on cmd/kubenab/admissions.go:
```diff
@@ -32,6 +50,14 @@ type patch struct {
 }
 
 func mutateAdmissionReviewHandler(w http.ResponseWriter, r *http.Request) {
+	httpRequestsTotal.With(prometheus.Labels{"api_endpoint": "metrics"}).Inc()
```
Shouldn't the `api_endpoint` here be set to `mutate` instead of `metrics`?

l0nax replied:
Oh :sweat_smile:

Commit:
Set correct metrics Label for mutate-Handler

rimusz reviewed (approved):
LGTM
